// @nrwl/web/src/utils/web.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPolyfillsPartial = exports.getStylesPartial = exports.getWebConfig = void 0;
// TODO @FrozenPandaz we should remove the following imports
const browser_1 = require("./third-party/cli-files/models/webpack-configs/browser");
const common_1 = require("./third-party/cli-files/models/webpack-configs/common");
const styles_1 = require("./third-party/cli-files/models/webpack-configs/styles");
const path_1 = require("path");
const normalize_1 = require("./normalize");
const typescript_1 = require("@nrwl/workspace/src/utilities/typescript");
const config_1 = require("./config");
const legacy_index_html_webpack_plugin_1 = require("./third-party/cli-files/plugins/legacy-index-html-webpack-plugin");
const index_html_webpack_plugin_1 = require("./third-party/cli-files/plugins/index-html-webpack-plugin");
const package_chunk_sort_1 = require("./third-party/cli-files/utilities/package-chunk-sort");
const typescript_2 = require("typescript");
const loader_utils_1 = require("loader-utils");
const postcssImports = require("postcss-import");
const path = require("path");
const sass_1 = require("./sass");
function getWebConfig(root, sourceRoot, options, esm, isScriptOptimizeOn, configuration) {
    // TODO(jack): Remove this in Nx 13 and go back to proper import
    const { webpackMerge } = require('../webpack/entry');
    const tsConfig = typescript_1.readTsConfig(options.tsConfig);
    if (isScriptOptimizeOn) {
        // Angular CLI uses an environment variable (NG_BUILD_DIFFERENTIAL_FULL)
        // to determine whether to use the scriptTargetOverride
        // or the tsConfig target
        // We want to force the target if overridden
        tsConfig.options.target = typescript_2.ScriptTarget.ES5;
    }
    const wco = {
        root,
        projectRoot: path_1.resolve(root, sourceRoot),
        buildOptions: normalize_1.convertBuildOptions(options),
        esm,
        console,
        tsConfig,
        tsConfigPath: options.tsConfig,
    };
    return webpackMerge([
        _getBaseWebpackPartial(options, esm, isScriptOptimizeOn, tsConfig.options.emitDecoratorMetadata, configuration),
        getPolyfillsPartial(options.polyfills, options.es2015Polyfills, esm, isScriptOptimizeOn),
        getStylesPartial(wco.root, wco.buildOptions, options.extractCss),
        getCommonPartial(wco),
        getBrowserPartial(wco, options, isScriptOptimizeOn),
    ]);
}
exports.getWebConfig = getWebConfig;
function getBrowserPartial(wco, options, isScriptOptimizeOn) {
    const config = browser_1.getBrowserConfig(wco);
    const { isWebpack5 } = require('../webpack/entry');
    if (!isScriptOptimizeOn) {
        const { deployUrl, subresourceIntegrity, scripts = [], styles = [], index, baseHref, } = options;
        config.plugins.push(isWebpack5
            ? new index_html_webpack_plugin_1.IndexHtmlWebpackPlugin({
                indexPath: path_1.resolve(wco.root, index),
                outputPath: path_1.basename(index),
                baseHref,
                entrypoints: package_chunk_sort_1.generateEntryPoints({ scripts, styles }),
                deployUrl,
                sri: subresourceIntegrity,
                moduleEntrypoints: [],
                noModuleEntrypoints: ['polyfills-es5'],
            })
            : new legacy_index_html_webpack_plugin_1.LegacyIndexHtmlWebpackPlugin({
                input: path_1.resolve(wco.root, index),
                output: path_1.basename(index),
                baseHref,
                entrypoints: package_chunk_sort_1.generateEntryPoints({ scripts, styles }),
                deployUrl,
                sri: subresourceIntegrity,
                noModuleEntrypoints: ['polyfills-es5'],
            }));
    }
    return config;
}
function _getBaseWebpackPartial(options, esm, isScriptOptimizeOn, emitDecoratorMetadata, configuration) {
    let partial = config_1.getBaseWebpackPartial(options, esm, isScriptOptimizeOn, emitDecoratorMetadata, configuration);
    delete partial.resolve.mainFields;
    return partial;
}
function getCommonPartial(wco) {
    const commonConfig = common_1.getCommonConfig(wco);
    delete commonConfig.entry;
    delete commonConfig.resolve.modules;
    delete commonConfig.resolve.extensions;
    delete commonConfig.output.path;
    delete commonConfig.module;
    return commonConfig;
}
function getStylesPartial(root, options, extractCss) {
    var _a, _b;
    // TODO(jack): Remove this in Nx 13 and go back to proper imports
    const { MiniCssExtractPlugin } = require('../webpack/entry');
    const includePaths = [];
    if (((_b = (_a = options === null || options === void 0 ? void 0 : options.stylePreprocessorOptions) === null || _a === void 0 ? void 0 : _a.includePaths) === null || _b === void 0 ? void 0 : _b.length) > 0) {
        options.stylePreprocessorOptions.includePaths.forEach((includePath) => includePaths.push(path.resolve(root, includePath)));
    }
    const partial = styles_1.getStylesConfig(root, options, includePaths);
    const rules = partial.module.rules.map((rule) => {
        if (!Array.isArray(rule.use)) {
            return rule;
        }
        rule.use = rule.use.map((loaderConfig) => {
            if (typeof loaderConfig === 'object' &&
                loaderConfig.loader === require.resolve('raw-loader')) {
                return {
                    loader: require.resolve('style-loader'),
                };
            }
            return loaderConfig;
        });
        return rule;
    });
    const loaderModulesOptions = {
        modules: {
            mode: 'local',
            getLocalIdent: getCSSModuleLocalIdent,
        },
        importLoaders: 1,
    };
    const commonLoaders = [
        {
            loader: extractCss
                ? MiniCssExtractPlugin.loader
                : require.resolve('style-loader'),
        },
        {
            loader: require.resolve('css-loader'),
            options: loaderModulesOptions,
        },
        {
            loader: require.resolve('postcss-loader'),
            options: {
                implementation: require('postcss'),
                postcssOptions: (loader) => ({
                    plugins: [
                        postcssImports({
                            addModulesDirectories: includePaths,
                            resolve: (url) => url.startsWith('~') ? url.substr(1) : url,
                            load: (filename) => {
                                return new Promise((resolve, reject) => {
                                    loader.fs.readFile(filename, (err, data) => {
                                        if (err) {
                                            reject(err);
                                            return;
                                        }
                                        const content = data.toString();
                                        resolve(content);
                                    });
                                });
                            },
                        }),
                    ],
                }),
            },
        },
    ];
    partial.module.rules = [
        {
            test: /\.css$|\.scss$|\.sass$|\.less$|\.styl$/,
            oneOf: [
                {
                    test: /\.module\.css$/,
                    use: commonLoaders,
                },
                {
                    test: /\.module\.(scss|sass)$/,
                    use: [
                        ...commonLoaders,
                        {
                            loader: require.resolve('sass-loader'),
                            options: {
                                implementation: sass_1.sassImplementation,
                                sassOptions: {
                                    fiber: false,
                                    precision: 8,
                                    includePaths,
                                },
                            },
                        },
                    ],
                },
                {
                    test: /\.module\.less$/,
                    use: [
                        ...commonLoaders,
                        {
                            loader: require.resolve('less-loader'),
                            options: {
                                paths: includePaths,
                            },
                        },
                    ],
                },
                {
                    test: /\.module\.styl$/,
                    use: [
                        ...commonLoaders,
                        {
                            loader: require.resolve('stylus-loader'),
                            options: {
                                paths: includePaths,
                            },
                        },
                    ],
                },
                ...rules,
            ],
        },
    ];
    return partial;
}
exports.getStylesPartial = getStylesPartial;
function getPolyfillsPartial(polyfills, es2015Polyfills, esm, isScriptOptimizeOn) {
    const config = {
        entry: {},
    };
    if (polyfills && esm && isScriptOptimizeOn) {
        // Safari 10.1 supports <script type="module"> but not <script nomodule>.
        // Need to patch it up so the browser doesn't load both sets.
        config.entry.polyfills = [
            require.resolve('@nrwl/web/src/utils/third-party/cli-files/models/safari-nomodule.js'),
            ...(polyfills ? [polyfills] : []),
        ];
    }
    else if (es2015Polyfills && !esm && isScriptOptimizeOn) {
        config.entry.polyfills = [
            es2015Polyfills,
            ...(polyfills ? [polyfills] : []),
        ];
    }
    else {
        if (polyfills) {
            config.entry.polyfills = [polyfills];
        }
        if (es2015Polyfills) {
            config.entry['polyfills-es5'] = [es2015Polyfills];
        }
    }
    return config;
}
exports.getPolyfillsPartial = getPolyfillsPartial;
function getCSSModuleLocalIdent(context, localIdentName, localName, options) {
    // Use the filename or folder name, based on some uses the index.js / index.module.(css|scss|sass) project style
    const fileNameOrFolder = context.resourcePath.match(/index\.module\.(css|scss|sass|styl)$/)
        ? '[folder]'
        : '[name]';
    // Create a hash based on a the file location and class name. Will be unique across a project, and close to globally unique.
    const hash = loader_utils_1.getHashDigest(path_1.posix.relative(context.rootContext, context.resourcePath) + localName, 'md5', 'base64', 5);
    // Use loaderUtils to find the file or folder name
    const className = loader_utils_1.interpolateName(context, `${fileNameOrFolder}_${localName}__${hash}`, options);
    // Remove the .module that appears in every classname when based on the file and replace all "." with "_".
    return className.replace('.module_', '_').replace(/\./g, '_');
}
