// @nrwl/workspace/src/command-line/clear-cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearCache = void 0;
const tslib_1 = require("tslib");
const app_root_1 = require("@nrwl/tao/src/utils/app-root");
const cache_directory_1 = require("../utilities/cache-directory");
const fs_extra_1 = require("fs-extra");
const output_1 = require("../utilities/output");
exports.clearCache = {
    command: 'clear-cache',
    describe: 'Clears all the cached Nx artifacts and metadata about the workspace.',
    handler: clearCacheHandler,
};
function clearCacheHandler() {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        output_1.output.note({
            title: 'Deleting the cache directory.',
            bodyLines: [`This might take a few minutes.`],
        });
        const dir = cache_directory_1.cacheDirectory(app_root_1.appRootPath, cache_directory_1.readCacheDirectoryProperty(app_root_1.appRootPath));
        fs_extra_1.removeSync(dir);
        output_1.output.success({
            title: 'Deleted the cache directory.',
        });
    });
}
