// @nrwl/workspace/src/command-line/connect-to-nx-cloud.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.connectToNxCloudCommand = exports.connectToNxCloudUsingScan = void 0;
const tslib_1 = require("tslib");
const file_utils_1 = require("../core/file-utils");
const output_1 = require("../utilities/output");
const devkit_1 = require("@nrwl/devkit");
const child_process_1 = require("child_process");
function connectToNxCloudUsingScan(scan) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (!scan)
            return;
        const nxJson = file_utils_1.readNxJson();
        const defaultRunnerIsUsed = Object.values(nxJson.tasksRunnerOptions).find((r) => r.runner == '@nrwl/workspace/tasks-runners/default');
        if (!defaultRunnerIsUsed)
            return;
        output_1.output.log({
            title: '--scan requires the workspace to be connected to Nx Cloud.',
        });
        const res = yield connectToNxCloudPrompt();
        if (res) {
            const pmc = devkit_1.getPackageManagerCommand();
            child_process_1.execSync(`${pmc.addDev} @nrwl/nx-cloud@latest`);
            child_process_1.execSync(`${pmc.exec} nx g @nrwl/nx-cloud:init`, {
                stdio: [0, 1, 2],
            });
        }
        else {
            output_1.output.log({ title: 'Executing the command without --scan' });
        }
    });
}
exports.connectToNxCloudUsingScan = connectToNxCloudUsingScan;
function connectToNxCloudCommand() {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const nxJson = file_utils_1.readNxJson();
        const nxCloudUsed = Object.values(nxJson.tasksRunnerOptions).find((r) => r.runner == '@nrwl/nx-cloud');
        if (nxCloudUsed) {
            output_1.output.log({
                title: 'This workspace is already connected to Nx Cloud.',
            });
            return;
        }
        const res = yield connectToNxCloudPrompt();
        if (!res)
            return;
        const pmc = devkit_1.getPackageManagerCommand();
        child_process_1.execSync(`${pmc.addDev} @nrwl/nx-cloud@latest`);
        child_process_1.execSync(`${pmc.exec} nx g @nrwl/nx-cloud:init`, {
            stdio: [0, 1, 2],
        });
    });
}
exports.connectToNxCloudCommand = connectToNxCloudCommand;
function connectToNxCloudPrompt() {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        return yield (yield Promise.resolve().then(() => require('enquirer')))
            .prompt([
            {
                name: 'NxCloud',
                message: `Connect to Nx Cloud? (It's free and doesn't require registration.)`,
                type: 'select',
                choices: [
                    {
                        name: 'Yes',
                        hint: 'Faster builds, run details, GitHub integration. Learn more at https://nx.app',
                    },
                    {
                        name: 'No',
                    },
                ],
                initial: 'No',
            },
        ])
            .then((a) => a.NxCloud === 'Yes');
    });
}
