// @nrwl/workspace/src/command-line/workspace-generators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.workspaceGenerators = void 0;
const tslib_1 = require("tslib");
const chalk = require("chalk");
const child_process_1 = require("child_process");
const fs_1 = require("fs");
const fs_extra_1 = require("fs-extra");
const path = require("path");
const yargsParser = require("yargs-parser");
const app_root_1 = require("@nrwl/tao/src/utils/app-root");
const fileutils_1 = require("../utilities/fileutils");
const output_1 = require("../utilities/output");
const devkit_1 = require("@nrwl/devkit");
const generate_1 = require("@nrwl/tao/src/commands/generate");
const rootDirectory = app_root_1.appRootPath;
const toolsDir = path.join(rootDirectory, 'tools');
const generatorsDir = path.join(toolsDir, 'generators');
const toolsTsConfigPath = path.join(toolsDir, 'tsconfig.tools.json');
function workspaceGenerators(args) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const outDir = compileTools();
        const parsedArgs = parseOptions(args, outDir);
        const collectionFile = path.join(outDir, 'workspace-generators.json');
        if (parsedArgs.listGenerators) {
            return listGenerators(collectionFile);
        }
        const generatorName = args[0];
        args[0] = collectionFile + ':' + generatorName;
        process.exitCode = yield generate_1.generate(process.cwd(), rootDirectory, args, parsedArgs.verbose);
    });
}
exports.workspaceGenerators = workspaceGenerators;
// compile tools
function compileTools() {
    const toolsOutDir = getToolsOutDir();
    fs_extra_1.removeSync(toolsOutDir);
    compileToolsDir(toolsOutDir);
    const generatorsOutDir = path.join(toolsOutDir, 'generators');
    const collectionData = constructCollection();
    devkit_1.writeJsonFile(path.join(generatorsOutDir, 'workspace-generators.json'), collectionData);
    return generatorsOutDir;
}
function getToolsOutDir() {
    return path.resolve(toolsDir, toolsTsConfig().compilerOptions.outDir);
}
function compileToolsDir(outDir) {
    fs_extra_1.copySync(generatorsDir, path.join(outDir, 'generators'));
    const tmpTsConfigPath = createTmpTsConfig(toolsTsConfigPath, {
        include: [path.join(generatorsDir, '**/*.ts')],
    });
    const pmc = devkit_1.getPackageManagerCommand();
    const tsc = `${pmc.exec} tsc`;
    try {
        child_process_1.execSync(`${tsc} -p ${tmpTsConfigPath}`, {
            stdio: 'inherit',
            cwd: rootDirectory,
        });
    }
    catch (_a) {
        process.exit(1);
    }
}
function constructCollection() {
    const generators = {};
    fs_1.readdirSync(generatorsDir).forEach((c) => {
        const childDir = path.join(generatorsDir, c);
        if (fs_1.existsSync(path.join(childDir, 'schema.json'))) {
            generators[c] = {
                factory: `./${c}`,
                schema: `./${devkit_1.normalizePath(path.join(c, 'schema.json'))}`,
                description: `Schematic ${c}`,
            };
        }
    });
    return {
        name: 'workspace-generators',
        version: '1.0',
        generators,
        schematics: generators,
    };
}
function toolsTsConfig() {
    return devkit_1.readJsonFile(toolsTsConfigPath);
}
function listGenerators(collectionFile) {
    try {
        const bodyLines = [];
        const collection = devkit_1.readJsonFile(collectionFile);
        bodyLines.push(chalk.bold(chalk.green('WORKSPACE GENERATORS')));
        bodyLines.push('');
        bodyLines.push(...Object.entries(collection.generators).map(([schematicName, schematicMeta]) => {
            return `${chalk.bold(schematicName)} : ${schematicMeta.description}`;
        }));
        bodyLines.push('');
        output_1.output.log({
            title: '',
            bodyLines,
        });
    }
    catch (error) {
        devkit_1.logger.fatal(error.message);
        return 1;
    }
    return 0;
}
function parseOptions(args, outDir) {
    const schemaPath = path.join(outDir, args[0], 'schema.json');
    let booleanProps = [];
    if (fileutils_1.fileExists(schemaPath)) {
        const { properties } = devkit_1.readJsonFile(path.join(outDir, args[0], 'schema.json'));
        if (properties) {
            booleanProps = Object.keys(properties).filter((key) => properties[key].type === 'boolean');
        }
    }
    return yargsParser(args, {
        boolean: ['dryRun', 'listGenerators', 'interactive', ...booleanProps],
        alias: {
            dryRun: ['d'],
            listSchematics: ['l'],
        },
        default: {
            interactive: true,
        },
    });
}
function createTmpTsConfig(tsconfigPath, updateConfig) {
    const tmpTsConfigPath = path.join(path.dirname(tsconfigPath), 'tsconfig.generated.json');
    const originalTSConfig = devkit_1.readJsonFile(tsconfigPath);
    const generatedTSConfig = Object.assign(Object.assign({}, originalTSConfig), updateConfig);
    process.on('exit', () => cleanupTmpTsConfigFile(tmpTsConfigPath));
    devkit_1.writeJsonFile(tmpTsConfigPath, generatedTSConfig);
    return tmpTsConfigPath;
}
function cleanupTmpTsConfigFile(tmpTsConfigPath) {
    if (tmpTsConfigPath) {
        fs_extra_1.removeSync(tmpTsConfigPath);
    }
}
