// @nrwl/workspace/src/command-line/workspace-results.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspaceResults = void 0;
const fileutils_1 = require("../utilities/fileutils");
const fs_1 = require("fs");
const fs_extra_1 = require("fs-extra");
const path_1 = require("path");
const app_root_1 = require("@nrwl/tao/src/utils/app-root");
const cache_directory_1 = require("../utilities/cache-directory");
const resultsDir = cache_directory_1.cacheDirectory(app_root_1.appRootPath, cache_directory_1.readCacheDirectoryProperty(app_root_1.appRootPath));
const resultsFile = path_1.join(resultsDir, 'results.json');
class WorkspaceResults {
    constructor(command, projects) {
        this.command = command;
        this.projects = projects;
        this.commandResults = {
            command: this.command,
            results: {},
        };
        this.startedWithFailedProjects = false;
        if (fs_1.existsSync(resultsFile)) {
            try {
                const commandResults = fileutils_1.readJsonFile(resultsFile);
                this.startedWithFailedProjects = commandResults.command === command;
                if (this.startedWithFailedProjects) {
                    this.commandResults = commandResults;
                    this.invalidateOldResults();
                }
            }
            catch (_a) {
                /**
                 * If we got here it is likely that RESULTS_FILE is not valid JSON.
                 * It is safe to continue, and it does not make much sense to give the
                 * user feedback as the file will be updated automatically.
                 */
            }
        }
    }
    get hasFailure() {
        return Object.values(this.commandResults.results).some((result) => !result);
    }
    getResult(projectName) {
        return this.commandResults.results[projectName];
    }
    saveResults() {
        fs_extra_1.ensureDirSync(resultsDir);
        if (this.hasFailure) {
            fileutils_1.writeJsonFile(resultsFile, this.commandResults);
        }
        else if (fs_1.existsSync(resultsFile)) {
            fs_1.unlinkSync(resultsFile);
        }
    }
    setResult(projectName, result) {
        this.commandResults.results[projectName] = result;
    }
    invalidateOldResults() {
        Object.keys(this.commandResults.results).forEach((projectName) => {
            if (!this.projects[projectName]) {
                delete this.commandResults.results[projectName];
            }
        });
    }
}
exports.WorkspaceResults = WorkspaceResults;
