// @nrwl/react/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.babelPluginStyledComponentsVersion = exports.eslintPluginReactHooksVersion = exports.eslintPluginReactVersion = exports.eslintPluginJsxA11yVersion = exports.eslintPluginImportVersion = exports.reactReduxVersion = exports.reduxjsToolkitVersion = exports.testingLibraryReactHooksVersion = exports.testingLibraryReactVersion = exports.typesReactRouterDomVersion = exports.reactRouterDomVersion = exports.typesStyledJsxVersion = exports.styledJsxVersion = exports.emotionBabelPlugin = exports.emotionReactVersion = exports.emotionStyledVersion = exports.typesStyledComponentsVersion = exports.styledComponentsVersion = exports.typesReactIsVersion = exports.typesReactDomVersion = exports.typesReactVersion = exports.reactIsVersion = exports.reactDomVersion = exports.reactVersion = exports.nxVersion = void 0;
exports.nxVersion = '12.10.0';
exports.reactVersion = '17.0.2';
exports.reactDomVersion = '17.0.2';
exports.reactIsVersion = '17.0.2';
exports.typesReactVersion = '17.0.3';
exports.typesReactDomVersion = '17.0.3';
exports.typesReactIsVersion = '17.0.0';
exports.styledComponentsVersion = '5.2.1';
exports.typesStyledComponentsVersion = '5.1.9';
exports.emotionStyledVersion = '11.1.5';
exports.emotionReactVersion = '11.1.5';
exports.emotionBabelPlugin = '11.2.0';
exports.styledJsxVersion = '3.4.4';
exports.typesStyledJsxVersion = '2.2.8';
exports.reactRouterDomVersion = '5.2.0';
exports.typesReactRouterDomVersion = '5.1.7';
exports.testingLibraryReactVersion = '11.2.6';
exports.testingLibraryReactHooksVersion = '7.0.1';
exports.reduxjsToolkitVersion = '1.5.1';
exports.reactReduxVersion = '7.2.3';
exports.eslintPluginImportVersion = '2.22.1';
exports.eslintPluginJsxA11yVersion = '6.4.1';
exports.eslintPluginReactVersion = '7.23.1';
exports.eslintPluginReactHooksVersion = '4.2.0';
exports.babelPluginStyledComponentsVersion = '1.10.7';
