// @nrwl/nx-cloud/lib/utilities/axios.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.axiosMultipleTries = exports.createApiAxiosInstance = exports.AxiosException = void 0;
const output_1 = require("@nrwl/workspace/src/utilities/output");
const environment_1 = require("./environment");
const waiter_1 = require("./waiter");
const axios = require('axios');
class AxiosException {
    constructor(type, message, axiosException) {
        this.type = type;
        this.message = message;
        this.axiosException = axiosException;
    }
}
exports.AxiosException = AxiosException;
function createApiAxiosInstance(options) {
    const baseUrl = options.url || 'https://api.nrwl.io';
    const authToken = environment_1.ACCESS_TOKEN ? environment_1.ACCESS_TOKEN : options.accessToken;
    if (!authToken) {
        throw new Error(`Unable to authenticate: ${authToken}`);
    }
    return axios.create({
        baseURL: baseUrl,
        timeout: environment_1.NX_CLOUD_NO_TIMEOUTS ? environment_1.UNLIMITED_TIMEOUT : 10000,
        headers: { authorization: authToken },
    });
}
exports.createApiAxiosInstance = createApiAxiosInstance;
function axiosMultipleTries(axiosCallCreator, attemptsLeft = 10) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            return yield axiosCallCreator();
        }
        catch (e) {
            const code = e.code || (e.response && e.response.status === 404 ? 404 : undefined);
            if (code === 'ECONNABORTED' ||
                code === 'ECONNREFUSED' ||
                code === 'EAI_AGAIN' ||
                code === 'ENOTFOUND' ||
                code === 'EPROTO' ||
                code === 404 ||
                code === 503) {
                if (attemptsLeft === 1) {
                    throw new AxiosException('timeout', `Cannot connect to Nx Cloud. Code: ${code}`, e);
                }
                else {
                    if (environment_1.VERBOSE_LOGGING) {
                        output_1.output.note({
                            title: `Received ${code}. Retrying in 10 seconds.`,
                        });
                    }
                    yield waiter_1.wait(10000);
                    return axiosMultipleTries(axiosCallCreator, attemptsLeft - 1);
                }
            }
            else {
                if (!e.response) {
                    throw new AxiosException('failure', `Error when connecting to Nx Cloud. Error: ${e.message}`, e);
                }
                let message = e.response.data.message
                    ? e.response.data.message
                    : e.response.data;
                if (typeof message !== 'string') {
                    message = e.message;
                }
                throw new AxiosException('failure', `Error when connecting to Nx Cloud. Error: ${message}`, e);
            }
        }
    });
}
exports.axiosMultipleTries = axiosMultipleTries;
