// @nrwl/workspace/src/utilities/typescript/compilation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.compileTypeScriptWatcher = exports.compileTypeScript = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const fs_extra_1 = require("fs-extra");
const ts = require("typescript");
const typescript_1 = require("../typescript");
function compileTypeScript(options) {
    const normalizedOptions = normalizeOptions(options);
    const tsConfig = getNormalizedTsConfig(normalizedOptions);
    if (normalizedOptions.deleteOutputPath) {
        fs_extra_1.removeSync(normalizedOptions.outputPath);
    }
    return createProgram(tsConfig, normalizedOptions.projectName);
}
exports.compileTypeScript = compileTypeScript;
function compileTypeScriptWatcher(options, callback) {
    const normalizedOptions = normalizeOptions(options);
    const tsConfig = getNormalizedTsConfig(normalizedOptions);
    if (normalizedOptions.deleteOutputPath) {
        fs_extra_1.removeSync(normalizedOptions.outputPath);
    }
    const host = ts.createWatchCompilerHost(tsConfig.fileNames, tsConfig.options, ts.sys);
    const original = host.onWatchStatusChange;
    host.onWatchStatusChange = (a, b, c, d) => tslib_1.__awaiter(this, void 0, void 0, function* () {
        original === null || original === void 0 ? void 0 : original(a, b, c, d);
        yield (callback === null || callback === void 0 ? void 0 : callback(a, b, c, d));
    });
    ts.createWatchProgram(host);
    return new Promise(() => { });
}
exports.compileTypeScriptWatcher = compileTypeScriptWatcher;
function getNormalizedTsConfig(options) {
    const tsConfig = typescript_1.readTsConfig(options.tsConfig);
    tsConfig.options.outDir = options.outputPath;
    tsConfig.options.noEmitOnError = true;
    tsConfig.options.rootDir = options.rootDir;
    return tsConfig;
}
function createProgram(tsconfig, projectName) {
    const host = ts.createCompilerHost(tsconfig.options);
    const program = ts.createProgram({
        rootNames: tsconfig.fileNames,
        options: tsconfig.options,
        host,
    });
    devkit_1.logger.info(`Compiling TypeScript files for project "${projectName}"...`);
    const results = program.emit();
    if (results.emitSkipped) {
        const diagnostics = ts.formatDiagnosticsWithColorAndContext(results.diagnostics, {
            getCurrentDirectory: () => ts.sys.getCurrentDirectory(),
            getNewLine: () => ts.sys.newLine,
            getCanonicalFileName: (name) => name,
        });
        devkit_1.logger.error(diagnostics);
        throw new Error(diagnostics);
    }
    else {
        devkit_1.logger.info(`Done compiling TypeScript files for project "${projectName}".`);
        return { success: true };
    }
}
function normalizeOptions(options) {
    var _a, _b;
    return Object.assign(Object.assign({}, options), { deleteOutputPath: (_a = options.deleteOutputPath) !== null && _a !== void 0 ? _a : true, rootDir: (_b = options.rootDir) !== null && _b !== void 0 ? _b : options.projectRoot });
}
