// @nrwl/workspace/src/core/nx-deps/nx-deps-cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractCachedFileData = exports.shouldRecomputeWholeGraph = exports.writeCache = exports.readCache = exports.ensureCacheDirectory = exports.nxDepsDir = void 0;
const path_1 = require("path");
const app_root_1 = require("@nrwl/tao/src/utils/app-root");
const fs_1 = require("fs");
const fs_extra_1 = require("fs-extra");
const fileutils_1 = require("../../utilities/fileutils");
const perf_hooks_1 = require("perf_hooks");
const cache_directory_1 = require("../../utilities/cache-directory");
exports.nxDepsDir = cache_directory_1.cacheDirectory(app_root_1.appRootPath, cache_directory_1.readCacheDirectoryProperty(app_root_1.appRootPath));
const nxDepsPath = path_1.join(exports.nxDepsDir, 'nxdeps.json');
function ensureCacheDirectory() {
    try {
        if (!fs_1.existsSync(exports.nxDepsDir)) {
            fs_extra_1.ensureDirSync(exports.nxDepsDir);
        }
    }
    catch (e) {
        /*
         * @jeffbcross: Node JS docs recommend against checking for existence of directory immediately before creating it.
         * Instead, just try to create the directory and handle the error.
         *
         * We ran into race conditions when running scripts concurrently, where multiple scripts were
         * arriving here simultaneously, checking for directory existence, then trying to create the directory simultaneously.
         *
         * In this case, we're creating the directory. If the operation failed, we ensure that the directory
         * exists before continuing (or raise an exception).
         */
        if (!fileutils_1.directoryExists(exports.nxDepsDir)) {
            throw new Error(`Failed to create directory: ${exports.nxDepsDir}`);
        }
    }
}
exports.ensureCacheDirectory = ensureCacheDirectory;
function readCache() {
    perf_hooks_1.performance.mark('read cache:start');
    ensureCacheDirectory();
    let data = null;
    try {
        if (fileutils_1.fileExists(nxDepsPath)) {
            data = fileutils_1.readJsonFile(nxDepsPath);
        }
    }
    catch (error) {
        console.log(`Error reading '${nxDepsPath}'. Continue the process without the cache.`);
        console.log(error);
    }
    perf_hooks_1.performance.mark('read cache:end');
    perf_hooks_1.performance.measure('read cache', 'read cache:start', 'read cache:end');
    return data !== null && data !== void 0 ? data : false;
}
exports.readCache = readCache;
function writeCache(packageJsonDeps, nxJson, tsConfig, projectGraph) {
    perf_hooks_1.performance.mark('write cache:start');
    const nxJsonPlugins = (nxJson.plugins || []).map((p) => ({
        name: p,
        version: packageJsonDeps[p],
    }));
    const newValue = {
        version: projectGraph.version || '4.0',
        deps: packageJsonDeps,
        pathMappings: tsConfig.compilerOptions.paths || {},
        nxJsonPlugins,
        nodes: projectGraph.nodes,
        dependencies: projectGraph.dependencies,
    };
    fileutils_1.writeJsonFile(nxDepsPath, newValue);
    perf_hooks_1.performance.mark('write cache:end');
    perf_hooks_1.performance.measure('write cache', 'write cache:start', 'write cache:end');
}
exports.writeCache = writeCache;
function shouldRecomputeWholeGraph(cache, packageJsonDeps, workspaceJson, nxJson, tsConfig) {
    if (cache.deps['@nrwl/workspace'] !== packageJsonDeps['@nrwl/workspace']) {
        return true;
    }
    // we have a cached project that is no longer present
    if (Object.keys(cache.nodes).some((p) => (cache.nodes[p].type === 'app' ||
        cache.nodes[p].type === 'lib' ||
        cache.nodes[p].type === 'e2e') &&
        !workspaceJson.projects[p])) {
        return true;
    }
    // a path mapping for an existing project has changed
    if (Object.keys(cache.pathMappings).some((t) => JSON.stringify(cache.pathMappings[t]) !=
        JSON.stringify(tsConfig.compilerOptions.paths[t]))) {
        return true;
    }
    // a new plugin has been added
    if ((nxJson.plugins || []).length !== cache.nxJsonPlugins.length)
        return true;
    // a plugin has changed
    if ((nxJson.plugins || []).some((t) => {
        const matchingPlugin = cache.nxJsonPlugins.find((p) => p.name === t);
        if (!matchingPlugin)
            return true;
        return matchingPlugin.version !== packageJsonDeps[t];
    })) {
        return true;
    }
    return false;
}
exports.shouldRecomputeWholeGraph = shouldRecomputeWholeGraph;
/*
This can only be invoked when the list of projects is either the same
or new projects have been added, so every project in the cache has a corresponding
project in fileMap
*/
function extractCachedFileData(fileMap, c) {
    const filesToProcess = {};
    const currentProjects = Object.keys(fileMap).filter((name) => fileMap[name].length > 0);
    const cachedFileData = {};
    currentProjects.forEach((p) => {
        processProjectNode(p, c.nodes[p], cachedFileData, filesToProcess, fileMap);
    });
    return {
        filesToProcess,
        cachedFileData,
    };
}
exports.extractCachedFileData = extractCachedFileData;
function processProjectNode(name, cachedNode, cachedFileData, filesToProcess, fileMap) {
    if (!cachedNode) {
        filesToProcess[name] = fileMap[name];
        return;
    }
    const fileDataFromCache = {};
    for (let f of cachedNode.data.files) {
        fileDataFromCache[f.file] = f;
    }
    if (!cachedFileData[name]) {
        cachedFileData[name] = {};
    }
    for (let f of fileMap[name]) {
        const fromCache = fileDataFromCache[f.file];
        if (fromCache && fromCache.hash == f.hash) {
            cachedFileData[name][f.file] = fromCache;
        }
        else {
            if (!filesToProcess[cachedNode.name]) {
                filesToProcess[cachedNode.name] = [];
            }
            filesToProcess[cachedNode.name].push(f);
        }
    }
}
