// @nrwl/workspace/src/core/hasher/hasher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hasher = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const workspace_1 = require("@nrwl/tao/src/shared/workspace");
const child_process_1 = require("child_process");
const fs_1 = require("fs");
const minimatch = require("minimatch");
const path_1 = require("path");
const perf_hooks_1 = require("perf_hooks");
const app_root_1 = require("../../utils/app-root");
const file_utils_1 = require("../file-utils");
const hashing_impl_1 = require("./hashing-impl");
const project_graph_1 = require("../project-graph");
class Hasher {
    constructor(projectGraph, nxJson, options, hashing = undefined) {
        var _a;
        this.projectGraph = projectGraph;
        this.nxJson = nxJson;
        this.options = options;
        if (!hashing) {
            this.hashing = hashing_impl_1.defaultHashing;
        }
        else {
            // this is only used for testing
            this.hashing = hashing;
        }
        this.projectHashes = new ProjectHasher(this.projectGraph, this.hashing, {
            selectivelyHashTsConfig: (_a = this.options.selectivelyHashTsConfig) !== null && _a !== void 0 ? _a : false,
        });
    }
    hashTaskWithDepsAndContext(task) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const command = this.hashCommand(task);
            const values = (yield Promise.all([
                this.projectHashes.hashProject(task.target.project, [
                    task.target.project,
                ]),
                this.implicitDepsHash(),
                this.runtimeInputsHash(),
            ]));
            const value = this.hashing.hashArray([
                Hasher.version,
                command,
                ...values.map((v) => v.value),
            ]);
            return {
                value,
                details: {
                    command,
                    nodes: values[0].nodes,
                    implicitDeps: values[1].files,
                    runtime: values[2].runtime,
                },
            };
        });
    }
    hashCommand(task) {
        var _a, _b, _c;
        return this.hashing.hashArray([
            (_a = task.target.project) !== null && _a !== void 0 ? _a : '',
            (_b = task.target.target) !== null && _b !== void 0 ? _b : '',
            (_c = task.target.configuration) !== null && _c !== void 0 ? _c : '',
            JSON.stringify(task.overrides),
        ]);
    }
    hashContext() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const values = (yield Promise.all([
                this.implicitDepsHash(),
                this.runtimeInputsHash(),
            ]));
            return {
                implicitDeps: values[0],
                runtime: values[1],
            };
        });
    }
    hashSource(task) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.projectHashes.hashProjectNodeSource(task.target.project);
        });
    }
    hashArray(values) {
        return this.hashing.hashArray(values);
    }
    runtimeInputsHash() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.runtimeInputs)
                return this.runtimeInputs;
            perf_hooks_1.performance.mark('hasher:runtime inputs hash:start');
            this.runtimeInputs = new Promise((res, rej) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                const inputs = this.options && this.options.runtimeCacheInputs
                    ? this.options.runtimeCacheInputs
                    : [];
                if (inputs.length > 0) {
                    try {
                        const values = (yield Promise.all(inputs.map((input) => new Promise((res, rej) => {
                            child_process_1.exec(input, (err, stdout, stderr) => {
                                if (err) {
                                    rej(err);
                                }
                                else {
                                    res({ input, value: `${stdout}${stderr}`.trim() });
                                }
                            });
                        }))));
                        const value = this.hashing.hashArray(values.map((v) => v.value));
                        const runtime = values.reduce((m, c) => ((m[c.input] = c.value), m), {});
                        perf_hooks_1.performance.mark('hasher:runtime inputs hash:end');
                        perf_hooks_1.performance.measure('hasher:runtime inputs hash', 'hasher:runtime inputs hash:start', 'hasher:runtime inputs hash:end');
                        res({ value, runtime });
                    }
                    catch (e) {
                        rej(new Error(`Nx failed to execute runtimeCacheInputs defined in nx.json failed:\n${e.message}`));
                    }
                }
                else {
                    res({ value: '', runtime: {} });
                }
            }));
            return this.runtimeInputs;
        });
    }
    implicitDepsHash() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.implicitDependencies)
                return this.implicitDependencies;
            perf_hooks_1.performance.mark('hasher:implicit deps hash:start');
            this.implicitDependencies = new Promise((res) => {
                var _a, _b;
                const implicitDeps = Object.keys((_a = this.nxJson.implicitDependencies) !== null && _a !== void 0 ? _a : {});
                const filesWithoutPatterns = implicitDeps.filter((p) => p.indexOf('*') === -1);
                const patterns = implicitDeps.filter((p) => p.indexOf('*') !== -1);
                const implicitDepsFromPatterns = patterns.length > 0
                    ? ((_b = this.projectGraph.allWorkspaceFiles) !== null && _b !== void 0 ? _b : [])
                        .filter((f) => !!patterns.find((pattern) => minimatch(f.file, pattern)))
                        .map((f) => f.file)
                    : [];
                const fileNames = [
                    ...filesWithoutPatterns,
                    ...implicitDepsFromPatterns,
                    //TODO: vsavkin move the special cases into explicit ts support
                    'package-lock.json',
                    'yarn.lock',
                    'pnpm-lock.yaml',
                    // ignore files will change the set of inputs to the hasher
                    '.gitignore',
                    '.nxignore',
                ];
                const fileHashes = [
                    ...fileNames
                        .map((maybeRelativePath) => {
                        // Normalize the path to always be absolute and starting with appRootPath so we can check it exists
                        if (!maybeRelativePath.startsWith(app_root_1.appRootPath)) {
                            return path_1.join(app_root_1.appRootPath, maybeRelativePath);
                        }
                        return maybeRelativePath;
                    })
                        .filter((file) => fs_1.existsSync(file))
                        .map((file) => {
                        const hash = this.hashing.hashFile(file);
                        return { file, hash };
                    }),
                    ...this.hashNxJson(),
                ];
                const combinedHash = this.hashing.hashArray(fileHashes.map((v) => v.hash));
                perf_hooks_1.performance.mark('hasher:implicit deps hash:end');
                perf_hooks_1.performance.measure('hasher:implicit deps hash', 'hasher:implicit deps hash:start', 'hasher:implicit deps hash:end');
                res({
                    value: combinedHash,
                    files: fileHashes.reduce((m, c) => ((m[c.file] = c.hash), m), {}),
                });
            });
            return this.implicitDependencies;
        });
    }
    hashNxJson() {
        const nxJsonPath = path_1.join(app_root_1.appRootPath, 'nx.json');
        if (!fs_1.existsSync(nxJsonPath)) {
            return [];
        }
        let nxJsonContents = '{}';
        try {
            const fileContents = fs_1.readFileSync(nxJsonPath, 'utf-8');
            const r = devkit_1.parseJson(fileContents);
            delete r.projects;
            nxJsonContents = JSON.stringify(r);
        }
        catch (_a) { }
        return [
            {
                hash: this.hashing.hashArray([nxJsonContents]),
                file: 'nx.json',
            },
        ];
    }
}
exports.Hasher = Hasher;
Hasher.version = '2.0';
class ProjectHasher {
    constructor(projectGraph, hashing, options) {
        this.projectGraph = projectGraph;
        this.hashing = hashing;
        this.options = options;
        this.sourceHashes = {};
        this.workspaceJson = this.readWorkspaceConfigFile(file_utils_1.workspaceFileName());
        this.nxJson = this.readNxJsonConfigFile('nx.json');
        this.tsConfigJson = this.readTsConfig();
    }
    hashProject(projectName, visited) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return Promise.resolve().then(() => tslib_1.__awaiter(this, void 0, void 0, function* () {
                var _a;
                const deps = (_a = this.projectGraph.dependencies[projectName]) !== null && _a !== void 0 ? _a : [];
                const depHashes = (yield Promise.all(deps.map((d) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                    if (visited.indexOf(d.target) > -1) {
                        return null;
                    }
                    else {
                        visited.push(d.target);
                        return yield this.hashProject(d.target, visited);
                    }
                })))).filter((r) => !!r);
                const projectHash = yield this.hashProjectNodeSource(projectName);
                const nodes = depHashes.reduce((m, c) => {
                    return Object.assign(Object.assign({}, m), c.nodes);
                }, { [projectName]: projectHash });
                const value = this.hashing.hashArray([
                    ...depHashes.map((d) => d.value),
                    projectHash,
                ]);
                return { value, nodes };
            }));
        });
    }
    hashProjectNodeSource(projectName) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.sourceHashes[projectName]) {
                this.sourceHashes[projectName] = new Promise((res) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                    var _a, _b;
                    const p = this.projectGraph.nodes[projectName];
                    if (project_graph_1.isNpmProject(p)) {
                        res(this.hashing.hashArray([p.data.version]));
                        return;
                    }
                    const fileNames = p.data.files.map((f) => f.file);
                    const values = p.data.files.map((f) => f.hash);
                    const workspaceJson = JSON.stringify((_a = this.workspaceJson.projects[projectName]) !== null && _a !== void 0 ? _a : '');
                    const nxJson = JSON.stringify((_b = this.nxJson.projects[projectName]) !== null && _b !== void 0 ? _b : '');
                    let tsConfig;
                    if (this.options.selectivelyHashTsConfig) {
                        tsConfig = this.removeOtherProjectsPathRecords(projectName);
                    }
                    else {
                        tsConfig = JSON.stringify(this.tsConfigJson);
                    }
                    res(this.hashing.hashArray([
                        ...fileNames,
                        ...values,
                        workspaceJson,
                        nxJson,
                        tsConfig,
                    ]));
                }));
            }
            return this.sourceHashes[projectName];
        });
    }
    removeOtherProjectsPathRecords(projectName) {
        var _a;
        const _b = this.tsConfigJson.compilerOptions, { paths } = _b, compilerOptions = tslib_1.__rest(_b, ["paths"]);
        const rootPath = this.workspaceJson.projects[projectName].root.split('/');
        rootPath.shift();
        const pathAlias = `@${this.nxJson.npmScope}/${rootPath.join('/')}`;
        return JSON.stringify({
            compilerOptions: Object.assign(Object.assign({}, compilerOptions), { paths: {
                    [pathAlias]: (_a = paths[pathAlias]) !== null && _a !== void 0 ? _a : [],
                } }),
        });
    }
    readTsConfig() {
        var _a;
        var _b;
        try {
            const res = devkit_1.readJsonFile('tsconfig.base.json');
            (_a = (_b = res.compilerOptions).paths) !== null && _a !== void 0 ? _a : (_b.paths = {});
            return res;
        }
        catch (_c) {
            return {
                compilerOptions: { paths: {} },
            };
        }
    }
    readWorkspaceConfigFile(path) {
        var _a;
        try {
            const res = devkit_1.readJsonFile(path);
            (_a = res.projects) !== null && _a !== void 0 ? _a : (res.projects = {});
            return workspace_1.resolveNewFormatWithInlineProjects(res);
        }
        catch (_b) {
            return { projects: {}, version: 2 };
        }
    }
    readNxJsonConfigFile(path) {
        var _a;
        try {
            const res = devkit_1.readJsonFile(path);
            (_a = res.projects) !== null && _a !== void 0 ? _a : (res.projects = {});
            return res;
        }
        catch (_b) {
            return { projects: {}, npmScope: '' };
        }
    }
}
