// @nrwl/web/src/utils/third-party/cli-files/utils/index-file/index-html-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexHtmlGenerator = void 0;
const tslib_1 = require("tslib");
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
const fs = require("fs");
const path_1 = require("path");
const augment_index_html_1 = require("./augment-index-html");
function stripBom(data) {
    return data.replace(/^\uFEFF/, '');
}
class IndexHtmlGenerator {
    constructor(options) {
        this.options = options;
        const extraPlugins = [];
        this.plugins = [
            augmentIndexHtmlPlugin(this),
            ...extraPlugins,
            postTransformPlugin(this),
        ];
    }
    process(options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            let content = stripBom(yield this.readIndex(this.options.indexPath));
            const warnings = [];
            const errors = [];
            for (const plugin of this.plugins) {
                const result = yield plugin(content, options);
                if (typeof result === 'string') {
                    content = result;
                }
                else {
                    content = result.content;
                    if (result.warnings.length) {
                        warnings.push(...result.warnings);
                    }
                    if (result.errors.length) {
                        errors.push(...result.errors);
                    }
                }
            }
            return {
                content,
                warnings,
                errors,
            };
        });
    }
    readAsset(path) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return fs.promises.readFile(path, 'utf-8');
        });
    }
    readIndex(path) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return fs.promises.readFile(path, 'utf-8');
        });
    }
}
exports.IndexHtmlGenerator = IndexHtmlGenerator;
function augmentIndexHtmlPlugin(generator) {
    const { deployUrl, crossOrigin, sri = false, entrypoints, } = generator.options;
    return (html, options) => tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { lang, baseHref, outputPath = '', noModuleFiles, files, moduleFiles, } = options;
        return augment_index_html_1.augmentIndexHtml({
            html,
            baseHref,
            deployUrl,
            crossOrigin,
            sri,
            lang,
            entrypoints,
            loadOutputFile: (filePath) => generator.readAsset(path_1.join(outputPath, filePath)),
            noModuleFiles,
            moduleFiles,
            files,
        });
    });
}
function postTransformPlugin({ options, }) {
    return (html) => tslib_1.__awaiter(this, void 0, void 0, function* () { return options.postTransform ? options.postTransform(html) : html; });
}
