// @nrwl/web/src/webpack/bundle5.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const devkit_1 = require("@nrwl/devkit");
const chalk = require("chalk");
const require_shim_1 = require("./require-shim");
const packageJson = require("../../package.json");
function validateVersion(path) {
    if (packageJson.dependencies[path] ===
        require_shim_1.requireShim(`${path}/package.json`).version) {
        devkit_1.logger.warn(`Found an outdated version of ${chalk.bold(path)}\n`);
        devkit_1.logger.info(devkit_1.stripIndents `
      If you want to use webpack 5, try installing compatible versions of the plugins.
      See: https://nx.dev/guides/webpack-5
    `);
        throw new Error('Incompatible version');
    }
}
module.exports = function (onFallback) {
    try {
        validateVersion('copy-webpack-plugin');
        validateVersion('terser-webpack-plugin');
        validateVersion('mini-css-extract-plugin');
        validateVersion('source-map-loader');
        validateVersion('webpack-merge');
        validateVersion('webpack-sources');
    }
    catch (_a) {
        devkit_1.logger.info(`NX Falling back to webpack 4 due to incompatible plugin versions`);
        onFallback();
        return require('./bundle4')();
    }
    return {
        CopyWebpackPlugin: require_shim_1.requireShim('copy-webpack-plugin'),
        MiniCssExtractPlugin: require_shim_1.requireShim('mini-css-extract-plugin'),
        SourceMapLoader: require_shim_1.requireShim('source-map-loader'),
        TerserWebpackPlugin: require_shim_1.requireShim('terser-webpack-plugin'),
        webpackSources: require_shim_1.requireShim('webpack-sources'),
        webpack: require_shim_1.requireShim('webpack'),
        webpackMerge: require_shim_1.requireShim('webpack-merge').default,
    };
};
