// @nrwl/react/plugins/storybook/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.webpack = exports.core = exports.babelDefault = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const app_root_1 = require("@nrwl/tao/src/utils/app-root");
const config_1 = require("@nrwl/web/src/utils/config");
const web_config_1 = require("@nrwl/web/src/utils/web.config");
const fileutils_1 = require("@nrwl/workspace/src/utilities/fileutils");
const version_utils_1 = require("@nrwl/workspace/src/utilities/version-utils");
const node_logger_1 = require("@storybook/node-logger");
const path_1 = require("path");
const semver_1 = require("semver");
const mergeWebpack = require("webpack-merge");
const merge_plugins_1 = require("./merge-plugins");
const reactWebpackConfig = require('../webpack');
const babelDefault = () => {
    // Add babel plugin for styled-components or emotion.
    // We don't have a good way to know when a project uses one or the other, so
    // add the plugin only if the other style package isn't used.
    const packageJson = fileutils_1.readJsonFile(path_1.join(app_root_1.appRootPath, 'package.json'));
    const deps = Object.assign(Object.assign({}, packageJson.dependencies), packageJson.devDependencies);
    const hasStyledComponents = !!deps['styled-components'];
    const plugins = [];
    if (hasStyledComponents) {
        plugins.push(['styled-components', { ssr: false }]);
    }
    return {
        presets: [],
        plugins: [...plugins],
    };
};
exports.babelDefault = babelDefault;
const core = (prev, options) => (Object.assign(Object.assign({}, prev), { disableWebpackDefaults: true }));
exports.core = core;
const webpack = (storybookWebpackConfig = {}, options) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    var _a, _b, _c, _d;
    node_logger_1.logger.info('=> Loading Nrwl React Storybook preset from "@nrwl/react/plugins/storybook"');
    const tsconfigPath = path_1.join(options.configDir, 'tsconfig.json');
    const builderOptions = Object.assign(Object.assign({}, options), { root: options.configDir, sourceRoot: '', fileReplacements: [], sourceMap: {
            hidden: false,
            scripts: true,
            styles: true,
            vendors: false,
        }, styles: [], optimization: {}, tsConfig: tsconfigPath, extractCss: storybookWebpackConfig.mode === 'production' });
    const esm = true;
    const isScriptOptimizeOn = storybookWebpackConfig.mode !== 'development';
    const extractCss = storybookWebpackConfig.mode === 'production';
    // ESM build for modern browsers.
    const baseWebpackConfig = mergeWebpack([
        config_1.getBaseWebpackPartial(builderOptions, esm, isScriptOptimizeOn),
        web_config_1.getStylesPartial(options.configDir, builderOptions, extractCss),
    ]);
    // run it through the React customizations
    const finalConfig = reactWebpackConfig(baseWebpackConfig);
    // Check whether the project .babelrc uses @emotion/babel-plugin. There's currently
    // a Storybook issue (https://github.com/storybookjs/storybook/issues/13277) which apparently
    // doesn't work with `@emotion/*` >= v11
    // this is a workaround to fix that
    let resolvedEmotionAliases = {};
    const packageJson = fileutils_1.readJsonFile(path_1.join(app_root_1.appRootPath, 'package.json'));
    const deps = Object.assign(Object.assign({}, packageJson.dependencies), packageJson.devDependencies);
    const emotionReactVersion = deps['@emotion/react'];
    if (emotionReactVersion &&
        semver_1.gte(version_utils_1.checkAndCleanWithSemver('@emotion/react', emotionReactVersion), '11.0.0')) {
        const babelrc = fileutils_1.readJsonFile(devkit_1.joinPathFragments(options.configDir, '../', '.babelrc'));
        if (babelrc.plugins.includes('@emotion/babel-plugin')) {
            resolvedEmotionAliases = {
                resolve: {
                    alias: {
                        '@emotion/core': devkit_1.joinPathFragments(app_root_1.appRootPath, 'node_modules', '@emotion/react'),
                        '@emotion/styled': devkit_1.joinPathFragments(app_root_1.appRootPath, 'node_modules', '@emotion/styled'),
                        'emotion-theming': devkit_1.joinPathFragments(app_root_1.appRootPath, 'node_modules', '@emotion/react'),
                    },
                },
            };
        }
    }
    return mergeWebpack(Object.assign(Object.assign({}, storybookWebpackConfig), { module: Object.assign(Object.assign({}, storybookWebpackConfig.module), { rules: [
                ...storybookWebpackConfig.module.rules,
                ...finalConfig.module.rules,
            ] }), resolve: Object.assign(Object.assign({}, storybookWebpackConfig.resolve), { plugins: merge_plugins_1.mergePlugins(...((_a = storybookWebpackConfig.resolve.plugins) !== null && _a !== void 0 ? _a : []), ...((_b = finalConfig.resolve.plugins) !== null && _b !== void 0 ? _b : [])) }), plugins: merge_plugins_1.mergePlugins(...((_c = storybookWebpackConfig.plugins) !== null && _c !== void 0 ? _c : []), ...((_d = finalConfig.plugins) !== null && _d !== void 0 ? _d : [])) }), resolvedEmotionAliases);
});
exports.webpack = webpack;
