// @nrwl/workspace/src/tasks-runner/cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cache = void 0;
const tslib_1 = require("tslib");
const app_root_1 = require("@nrwl/tao/src/utils/app-root");
const fs_1 = require("fs");
const fs_extra_1 = require("fs-extra");
const path_1 = require("path");
const child_process_1 = require("child_process");
const cache_directory_1 = require("../utilities/cache-directory");
const util = require('util');
const readFileAsync = util.promisify(fs_extra_1.readFile);
const existsAsync = util.promisify(fs_1.exists);
const lstatAsync = util.promisify(fs_1.lstat);
const readdirAsync = util.promisify(fs_1.readdir);
class Cache {
    constructor(options) {
        this.options = options;
        this.root = app_root_1.appRootPath;
        this.cachePath = this.createCacheDir();
        this.terminalOutputsDir = this.createTerminalOutputsDir();
        this.latestOutputsHashesDir = this.ensureLatestOutputsHashesDir();
    }
    removeOldCacheRecords() {
        /**
         * Even though spawning a process is fast, we don't want to do it every time
         * the user runs a command. Instead, we want to do it once in a while.
         */
        const shouldSpawnProcess = Math.floor(Math.random() * 50) === 1;
        if (shouldSpawnProcess) {
            const scriptPath = require.resolve('@nrwl/workspace/src/tasks-runner/remove-old-cache-records.js', { paths: [this.root] });
            try {
                const p = child_process_1.spawn('node', [scriptPath, `"${this.cachePath}"`], {
                    stdio: 'ignore',
                    detached: true,
                });
                p.unref();
            }
            catch (e) {
                console.log(`Unable to start remove-old-cache-records script:`);
                console.log(e.message);
            }
        }
    }
    get(task) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const res = yield this.getFromLocalDir(task);
            // didn't find it locally but we have a remote cache
            if (!res && this.options.remoteCache) {
                // attempt remote cache
                yield this.options.remoteCache.retrieve(task.hash, this.cachePath);
                // try again from local cache
                return this.getFromLocalDir(task);
            }
            else {
                return res;
            }
        });
    }
    put(task, terminalOutput, outputs, code) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const td = path_1.join(this.cachePath, task.hash);
            const tdCommit = path_1.join(this.cachePath, `${task.hash}.commit`);
            // might be left overs from partially-completed cache invocations
            yield fs_extra_1.remove(tdCommit);
            yield fs_extra_1.remove(td);
            yield fs_extra_1.mkdir(td);
            yield fs_extra_1.writeFile(path_1.join(td, 'terminalOutput'), terminalOutput !== null && terminalOutput !== void 0 ? terminalOutput : 'no terminal output');
            yield fs_extra_1.mkdir(path_1.join(td, 'outputs'));
            for (const f of outputs) {
                const src = path_1.join(this.root, f);
                if (yield existsAsync(src)) {
                    const cached = path_1.join(td, 'outputs', f);
                    // Ensure parent directory is created if src is a file
                    const isFile = (yield lstatAsync(src)).isFile();
                    const directory = isFile ? path_1.resolve(cached, '..') : cached;
                    yield fs_extra_1.ensureDir(directory);
                    yield fs_extra_1.copy(src, cached);
                }
            }
            // we need this file to account for partial writes to the cache folder.
            // creating this file is atomic, whereas creating a folder is not.
            // so if the process gets terminated while we are copying stuff into cache,
            // the cache entry won't be used.
            yield fs_extra_1.writeFile(path_1.join(td, 'code'), code.toString());
            yield fs_extra_1.writeFile(tdCommit, 'true');
            if (this.options.remoteCache) {
                yield this.options.remoteCache.store(task.hash, this.cachePath);
            }
        });
    }
    copyFilesFromCache(hash, cachedResult, outputs) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.removeRecordedOutputsHashes(outputs);
            for (const f of outputs) {
                const cached = path_1.join(cachedResult.outputsPath, f);
                if (yield existsAsync(cached)) {
                    const isFile = (yield lstatAsync(cached)).isFile();
                    const src = path_1.join(this.root, f);
                    yield fs_extra_1.remove(src);
                    // Ensure parent directory is created if src is a file
                    const directory = isFile ? path_1.resolve(src, '..') : src;
                    yield fs_extra_1.ensureDir(directory);
                    yield fs_extra_1.copy(cached, src);
                }
            }
            yield this.recordOutputsHash(outputs, hash);
        });
    }
    temporaryOutputPath(task) {
        return path_1.join(this.terminalOutputsDir, task.hash);
    }
    removeRecordedOutputsHashes(outputs) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            for (const output of outputs) {
                const hashFile = this.getFileNameWithLatestRecordedHashForOutput(output);
                try {
                    yield fs_extra_1.unlink(hashFile);
                }
                catch (e) { }
            }
        });
    }
    recordOutputsHash(outputs, hash) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            for (const output of outputs) {
                const hashFile = this.getFileNameWithLatestRecordedHashForOutput(output);
                try {
                    yield fs_extra_1.ensureDir(path_1.dirname(hashFile));
                    yield fs_extra_1.writeFile(hashFile, hash);
                }
                catch (_a) { }
            }
        });
    }
    shouldCopyOutputsFromCache(taskWithCachedResult, outputs) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return ((yield this.areLatestOutputsHashesDifferentThanTaskHash(outputs, taskWithCachedResult.task.hash)) ||
                (yield this.isAnyOutputMissing(taskWithCachedResult.cachedResult, outputs)));
        });
    }
    areLatestOutputsHashesDifferentThanTaskHash(outputs, hash) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            for (let output of outputs) {
                if ((yield this.getLatestRecordedHashForTask(output)) !== hash)
                    return true;
            }
            return false;
        });
    }
    getLatestRecordedHashForTask(output) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                return (yield readFileAsync(this.getFileNameWithLatestRecordedHashForOutput(output))).toString();
            }
            catch (e) {
                return null;
            }
        });
    }
    isAnyOutputMissing(cachedResult, outputs) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            for (let output of outputs) {
                const cacheOutputPath = path_1.join(cachedResult.outputsPath, output);
                const rootOutputPath = path_1.join(this.root, output);
                if ((yield existsAsync(cacheOutputPath)) &&
                    (yield lstatAsync(cacheOutputPath)).isFile()) {
                    if ((yield existsAsync(path_1.join(cachedResult.outputsPath, output))) &&
                        !(yield existsAsync(path_1.join(this.root, output)))) {
                        return true;
                    }
                }
                const haveDifferentAmountOfFiles = (yield existsAsync(cacheOutputPath)) &&
                    (yield existsAsync(rootOutputPath)) &&
                    (yield readdirAsync(cacheOutputPath)).length !==
                        (yield readdirAsync(rootOutputPath)).length;
                if (((yield existsAsync(cacheOutputPath)) &&
                    !(yield existsAsync(rootOutputPath))) ||
                    haveDifferentAmountOfFiles) {
                    return true;
                }
            }
            return false;
        });
    }
    getFileNameWithLatestRecordedHashForOutput(output) {
        return path_1.join(this.latestOutputsHashesDir, `${output.split(path_1.sep).join('-')}.hash`);
    }
    getFromLocalDir(task) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const tdCommit = path_1.join(this.cachePath, `${task.hash}.commit`);
            const td = path_1.join(this.cachePath, task.hash);
            if (yield existsAsync(tdCommit)) {
                const terminalOutput = yield fs_extra_1.readFile(path_1.join(td, 'terminalOutput'), 'utf-8');
                let code = 0;
                try {
                    code = Number(yield fs_extra_1.readFile(path_1.join(td, 'code'), 'utf-8'));
                }
                catch (e) { }
                return {
                    terminalOutput,
                    outputsPath: path_1.join(td, 'outputs'),
                    code,
                };
            }
            else {
                return null;
            }
        });
    }
    createCacheDir() {
        const dir = cache_directory_1.cacheDirectory(this.root, this.options.cacheDirectory);
        fs_extra_1.ensureDirSync(dir);
        return dir;
    }
    createTerminalOutputsDir() {
        const path = path_1.join(this.cachePath, 'terminalOutputs');
        fs_extra_1.ensureDirSync(path);
        return path;
    }
    ensureLatestOutputsHashesDir() {
        const path = path_1.join(this.cachePath, 'latestOutputsHashes');
        fs_extra_1.ensureDirSync(path);
        return path;
    }
}
exports.Cache = Cache;
