// @nrwl/workspace/src/generators/npm-package/npm-package.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.npmPackageSchematic = exports.npmPackageGenerator = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const path_1 = require("path");
function normalizeOptions(options) {
    options.name = devkit_1.names(options.name).fileName;
    return options;
}
function addFiles(projectRoot, tree, npmScope, options) {
    const packageJsonPath = path_1.join(projectRoot, 'package.json');
    devkit_1.writeJson(tree, packageJsonPath, {
        name: path_1.join(`@${npmScope}`, options.name),
        scripts: {
            test: 'node index.js',
        },
    });
    devkit_1.generateFiles(tree, path_1.join(__dirname, './files'), projectRoot, {});
}
function npmPackageGenerator(tree, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        options = normalizeOptions(options);
        const { libsDir, npmScope } = devkit_1.getWorkspaceLayout(tree);
        const projectRoot = path_1.join(libsDir, options.name);
        devkit_1.addProjectConfiguration(tree, options.name, {
            root: projectRoot,
        });
        const isEmpty = tree.children(projectRoot).length === 0;
        if (isEmpty) {
            addFiles(projectRoot, tree, npmScope, options);
        }
        yield devkit_1.formatFiles(tree);
    });
}
exports.npmPackageGenerator = npmPackageGenerator;
exports.npmPackageSchematic = devkit_1.convertNxGenerator(npmPackageGenerator);
