// @nrwl/workspace/src/core/project-graph/build-nodes/workspace-projects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildWorkspaceProjectNodes = exports.convertNpmScriptsToTargets = void 0;
const file_utils_1 = require("../../file-utils");
function convertNpmScriptsToTargets(projectRoot) {
    try {
        const packageJsonString = file_utils_1.defaultFileRead(`${projectRoot}/package.json`).toString();
        const parsedPackagedJson = JSON.parse(packageJsonString);
        const res = {};
        // handle no scripts
        Object.keys(parsedPackagedJson.scripts || {}).forEach((script) => {
            res[script] = {
                executor: '@nrwl/workspace:run-script',
                options: {
                    script,
                },
            };
        });
        return res;
    }
    catch (e) {
        return undefined;
    }
}
exports.convertNpmScriptsToTargets = convertNpmScriptsToTargets;
function buildWorkspaceProjectNodes(ctx, builder) {
    const toAdd = [];
    Object.keys(ctx.workspace.projects).forEach((key) => {
        const p = ctx.workspace.projects[key];
        if (!p.targets) {
            p.targets = convertNpmScriptsToTargets(p.root);
        }
        const projectType = p.projectType === 'application'
            ? key.endsWith('-e2e')
                ? 'e2e'
                : 'app'
            : 'lib';
        const tags = ctx.workspace.projects && ctx.workspace.projects[key]
            ? ctx.workspace.projects[key].tags || []
            : [];
        toAdd.push({
            name: key,
            type: projectType,
            data: Object.assign(Object.assign({}, p), { tags, files: ctx.fileMap[key] }),
        });
    });
    // Sort by root directory length (do we need this?)
    toAdd.sort((a, b) => {
        if (!a.data.root)
            return -1;
        if (!b.data.root)
            return -1;
        return a.data.root.length > b.data.root.length ? -1 : 1;
    });
    toAdd.forEach((n) => {
        builder.addNode({
            name: n.name,
            type: n.type,
            data: n.data,
        });
    });
}
exports.buildWorkspaceProjectNodes = buildWorkspaceProjectNodes;
